### ISSUE TITLE
Async support

### ISSUE THREAD

**Issue by ramnes** [DonDebonair/slack-machine#591]
Hey there!

Do you have any plan for supporting async code?

I've seen that you want to use the Events API (#469), so maybe [bolt-python](https://github.com/slackapi/bolt-python) and its async backend could be used as a base?

**Comment by DonDebonair** [comment]
Hey there! How serendipitous that you ask about this! I just got back to Slack Machine yesterday after a hiatus of a couple of months. I started experimenting with async yesterday. My plan is not only to support async, but to make Slack Machine async-only.

My initial idea was to use Bolt, but I think it's overkill for the low-level access I need. Slack SDK for Python actually also supports SocketMode and it can work with the Events API just as well.

I also debated for a while if Bolt was making Slack Machine obsolete, but from what I've seen, Bolt doesn't really let you build Slack bots in a scalable way, using a plugin-based architecture (see [this issue](https://github.com/slackapi/bolt-python/issues/236) for the problems with Bolt). This is what Slack Machine does well. So if I can implement async support with SocketMode and the Event API, I think it could really help adoption of Slack Machine

Anyways, thanks for your interest! I hope to build a first async version this weekend and release it as an alpha

**Comment by DonDebonair** [comment]
Thanks for your elaborate answer! Very insightful!

With regards to the underlying base of Slack Machine, there are basically 2 options:

- [Bolt for Python](https://github.com/slackapi/bolt-python)
- [Python Slack SDK](https://github.com/slackapi/python-slack-sdk)

I don't think the latter will receive less engineering effort from the Slack team than the former, because Bolt is actually built on top of the Python Slack SDK. The Python Slack SDK is just more low-level than Bolt. And what Bolt adds on top of the Python Slack SDK, is mostly convenience in a similar way as Slack Machine does. E.g. decorators for writing handlers for certain events.

But Slack Machine augments that with a true plugin system, that should make it easy to distribute functionality separate from your bot instance and reuse it between bots.

When I look at the code examples between [Bolt](https://slack.dev/bolt-python/concepts#socket-mode) and the [Python Slack SDK](https://slack.dev/python-slack-sdk/socket-mode/index.html#asyncio-based-libraries), for my purposes there are not so many differences when it comes to using Socket Mode (which is what I'm focussing on right now).

So unless I'm missing something, there is no convincing reason to use Bolt instead of the Python Slack SDK, given that the latter will be maintained because it's used in the former?

With regards to the rule matching: when you talk about having typed arguments in your function and casting input automatically, how far would that go? What I mean is: FastAPI lets you define nested models (using Pydantic), which makes sense for an API with JSON input. For a Slack bot it's much simpeler, because you're always dealing with unstructured text (Slack messages) so the only casting you could do, is to `int`, `bool`, `str` and maybe `date`/`datetime`? Or am I missing something?

And you'd still need regex I think, because the text is unstructured?

Can you given an example of what a listener function would look like and what input it would respond to?

**Comment by DonDebonair** [comment]
Btw, I have a Gitter room setup (see README) to discuss things in chat, but it's not really used right now. We could use it to have more real-time in-depth discussions at some point. I also contemplated moving to Slack for the community.

**Comment by ramnes** [comment]
> I don't think the latter will receive less engineering effort from the Slack team than the former, because Bolt is actually built on top of the Python Slack SDK. The Python Slack SDK is just more low-level than Bolt.

> So unless I'm missing something, there is no convincing reason to use Bolt instead of the Python Slack SDK, given that the latter will be maintained because it's used in the former?

You're completely right. I was under the impression than Bolt was implementing most things async and events-related, but digging a bit more I'm discovering now that I completely overlooked the SDK, and that it's far more complete than I thought it was. Sorry for the noise!

> With regards to the rule matching: when you talk about having typed arguments in your function and casting input automatically, how far would that go? What I mean is: FastAPI lets you define nested models (using Pydantic), which makes sense for an API with JSON input. For a Slack bot it's much simpeler, because you're always dealing with unstructured text (Slack messages) so the only casting you could do, is to int, bool, str and maybe date/datetime? Or am I missing something?

I was referring to the [path parameters](https://fastapi.tiangolo.com/tutorial/path-params/). :)

An alternative is Flask / Starlette "convertors", that work with regular expressions underneath, but they don't rely on type hints and thus feel a bit less modern.

**Comment by DonDebonair** [comment]
So, I made quite some progress on the async story! You can find my work [in this draft PR](https://github.com/DonDebonair/slack-machine/pull/592). It's still WIP, but the basics work now:

- The core is completely async, based on SocketMode + events
- The plugin system is now async, so all plugin methods should be async as well
- The built-in plugins have been converted to async as well
- I've refactored a lot of stuff, adding type hints wherever I could

TODO:
- Currently, only the `@respond_to` and `@listen_to` decorators work in plugins. `@process` is next on my todo list and should be really easy.
- The RBAC plugin doesn't work yet
- The scheduler has been removed completely. I'm not sure if I'll bring it back. Removing the scheduler allowed me to get rid of the Singleton pattern that was plaguing the code. And I'm not sure if anyone was using it or not. If there is interest to bring it back, I can! But I'll have to rethink how I can persist scheduled tasks
- The exact type-matching for parameters extracted from the message (which you mentioned above) is something I'd like to do, but I'll have to look into how that would work
- Update documentation

The work on the branch can already be tested. I've added a script in the root called `run_dev_async.py` which runs the async version of Slack Machine. The code for sync + async currently live side by side, with the async code living in `machine_v2`. I want to finish this and start testing it rigorously. If it works, and some users (like you?) have tested it as well, I actually want to remove the sync code, because I don't plan on maintaining 2 versions. People can always fall back to older versions if they want to.

If you want to test it, you'll have to create your own Slack app. You can use the following App Manifest to get started quickly:

<details>
<summary>App Manifest</summary>
```yml
display_information:
  name: Slack Machine
features:
  bot_user:
    display_name: Slack Machine
    always_online: false
oauth_config:
  scopes:
    bot:
      - app_mentions:read
      - channels:history
      - channels:join
      - channels:read
      - chat:write
      - chat:write.public
      - emoji:read
      - groups:history
      - groups:read
      - groups:write
      - im:history
      - im:read
      - im:write
      - mpim:history
      - mpim:read
      - mpim:write
      - pins:read
      - pins:write
      - reactions:read
      - reactions:write
      - users:read
      - users:read.email
      - channels:manage
      - chat:write.customize
      - commands
      - dnd:read
      - files:read
      - files:write
      - links:read
      - links:write
      - metadata.message:read
      - usergroups:read
      - usergroups:write
      - users.profile:read
      - users:write
settings:
  event_subscriptions:
    bot_events:
      - app_mention
      - channel_archive
      - channel_created
      - channel_deleted
      - channel_id_changed
      - channel_left
      - channel_rename
      - channel_unarchive
      - group_archive
      - group_deleted
      - group_left
      - group_rename
      - group_unarchive
      - member_joined_channel
      - member_left_channel
      - message.channels
      - message.groups
      - message.im
      - message.mpim
      - reaction_added
      - reaction_removed
      - team_join
      - user_change
      - user_profile_changed
      - user_status_changed
  interactivity:
    is_enabled: true
  org_deploy_enabled: false
  socket_mode_enabled: true
  token_rotation_enabled: false
```
</details>

**Comment by DonDebonair** [comment]
Update `@process` now works as well!

Async code has been moved to the `machine.asyncio` package

**Comment by DonDebonair** [comment]
Update: RBAC now works! And all of the async code is type checked with MyPy

**Comment by DonDebonair** [comment]
This has been fixed now that #592 has been merged. Published as [v0.26.0](https://pypi.org/project/slack-machine/0.26.0/)
